cmd/pulumi-test-language/interface.go

```diff
@@ -918,6 +919,100 @@ func (eng *languageTestServer) RunLanguageTest(
 			}
 		}
 
+		// Query the language plugin for what it thinks the project packages are, we expect to see the SDKs.
+		packages, err := languageClient.GetRequiredPackages(programInfo)
+		if err != nil {
+			return makeTestResponse(fmt.Sprintf("get required packages: %v", err)), nil
+		}
+		expectedPackages := []workspace.PackageDescriptor{}
+		for _, pkg := range programPackages {
+			if pkg.Name() == "pulumi" {
+				// Skip the pulumi package, the version for that is handled above.
+				continue
+			}
+
+			pkgDef, err := pkg.Definition()
+			if err != nil {
+				return makeTestResponse(fmt.Sprintf("get package definition: %v", err)), nil
+			}
+
+			var desc workspace.PackageDescriptor
+			if pkgDef.Parameterization == nil {
+				desc = workspace.PackageDescriptor{
+					PluginSpec: workspace.PluginSpec{
+						Name:    pkgDef.Name,
+						Version: pkgDef.Version,
+					},
+				}
+			} else {
+				desc = workspace.PackageDescriptor{
+					PluginSpec: workspace.PluginSpec{
+						Name:    pkgDef.Parameterization.BaseProvider.Name,
+						Version: &pkgDef.Parameterization.BaseProvider.Version,
+					},
+					Parameterization: &workspace.Parameterization{
+						Name:    pkgDef.Name,
+						Version: *pkgDef.Version,
+						Value:   pkgDef.Parameterization.Parameter,
+					},
+				}
+			}
+
+			expectedPackages = append(expectedPackages, desc)
+		}
+
+		versionsMatch := func(expected, actual *semver.Version) bool {
+			if expected == nil && actual == nil {
+				return true
+			}
+			if expected == nil || actual == nil {
+				return false
+			}
+			return expected.EQ(*actual)
+		}
```

Comment:
Is this the same logic as the current/previous check? I vaguely recall in Java land omitting versions so that things would pass since there wasn't a reliable way to do it at the time, but this logic looks like if e.g. expected is not nil, but actual is (like in Java), we'll fail?

Reply:
This is for _packages_ not dependencies. We should be ok with dependencies not matching versions exactly, but packages should be the fully correct versions, because it's going to match what we'd send for RegisterResource requests.

Reply:
Makes sense; thanks!